### PR TITLE
Prisoner feature bugfix

### LINKED ISSUE #43: Rework or Revert r9095 - if POW, any mercs above 3 are removed from game permanent (not avoidable if more than 3 mercs)
Having Tixa as additional place for POW happening after we surrender our mercs is nice, but this part sucks big:

"fix: all mercs in the map will be taken POW when surrendered, however only 3 will appear in Alma/Tixa/N7, rest will be POW forever without chance of getting free"

Why doing this? Who would use the surrender if that is happening? The escape has been a fun side-quest, but when it results in mercs being lost for good, without a chance to free them, why would anybody use this?

I tried to revert this in my local copy, but it seems there are commits before (and maybe also after) that deal with this and I can't find out which files apart from the one at r9095 need to be reverted (comments on this are not always helpful, seem to miss some stuff).
I also didn't understand why it was nescessary to make this more-than-three cap, beyond my skill to find that out.

Like I said, it's a nice addition, but loosing mercs permanently without a chance is just bad, imo.

------

Revision: 9095
Author: Shadooow
Date: Montag, 21. Juni 2021 15:53:56
Message:
Major rework of getting captured quests and mechanics.
- new feature: added Tixa as another prison where capture mercs can be held (Alma is still primary choice though)
- fix: all mercs in the map will be taken POW when surrendered, however only 3 will appear in Alma/Tixa/N7, rest will be POW forever without chance of getting free
- fix: whenever player liberates any of the POW sectors (Alma, Tixa, N7) the respective quest/cutscene with POW will be disabled in that sector until enemy recapture the sector
- robot will be destroyed when surrendered as it makes no sense to inprison mechanical units
- externalized the GridNos for freeing POW mercs into Mod_Settings.ini
----
Modified : /trunk/GameSource/ja2_v1.13/Build/GameSettings.cpp
Modified : /trunk/GameSource/ja2_v1.13/Build/GameSettings.h
Modified : /trunk/GameSource/ja2_v1.13/Build/Strategic/Player Command.cpp
Modified : /trunk/GameSource/ja2_v1.13/Build/Strategic/Queen Command.cpp
Modified : /trunk/GameSource/ja2_v1.13/Build/Strategic/Quests.cpp
Modified : /trunk/GameSource/ja2_v1.13/Build/Strategic/Quests.h
Modified : /trunk/GameSource/ja2_v1.13/Build/Strategic/strategicmap.cpp
Modified : /trunk/GameSource/ja2_v1.13/Build/Tactical/Handle Doors.cpp
Modified : /trunk/GameSource/ja2_v1.13/Build/Tactical/Overhead.cpp
Modified : /trunk/GameSource/ja2_v1.13/Build/TacticalAI/DecideAction.cpp
Modified : /trunk/GameSource/ja2_v1.13/Build/TileEngine/Explosion Control.cpp

### PR DESCRIPTION
**Pull request by Asdow**
Will close #43 and #56 

Mercs that are not captured by the enemy when surrendering have a chance to escape to fight another day. Any captured POW can be rescued from the prison they got sent to.

### CODE REVIEW

File: Tactical/Overhead.cpp [hunk 1]
```diff
@@ -11092,6 +11034,105 @@ void HandleTurncoatAttempt( SOLDIERTYPE* pSoldier )
 	}
 }
 
+void EscapeTimerCallback()
+{
+    const bool chanceToEscape = Chance(75);
```
Comment (majcosta) [inline Tactical/Overhead.cpp:11039]: this can be static

File: Tactical/Overhead.cpp [hunk 2]
```diff
@@ -11092,6 +11034,105 @@ void HandleTurncoatAttempt( SOLDIERTYPE* pSoldier )
 	}
 }
 
+void EscapeTimerCallback()
+{
+    const bool chanceToEscape = Chance(75);
+    bool escaped = false;
+    // Look for an escape direction for remaining mercs
+    std::vector<WorldDirections> possibleEscapeDirections{ NORTH, EAST, SOUTH, WEST };
```
Comment (majcosta) [inline Tactical/Overhead.cpp:11042]: use a const std::array, they're not inventing new directions afaik

File: Tactical/Overhead.cpp [hunk 3]
```diff
@@ -11092,6 +11034,105 @@ void HandleTurncoatAttempt( SOLDIERTYPE* pSoldier )
 	}
 }
 
+void EscapeTimerCallback()
+{
+    const bool chanceToEscape = Chance(75);
+    bool escaped = false;
+    // Look for an escape direction for remaining mercs
+    std::vector<WorldDirections> possibleEscapeDirections{ NORTH, EAST, SOUTH, WEST };
+    std::random_shuffle(possibleEscapeDirections.begin(), possibleEscapeDirections.end());
```
Comment (majcosta) [inline Tactical/Overhead.cpp:11043]: random_shuffle is deprecated, use std::shuffle with a std::mt19937 generator instead

File: Tactical/Overhead.cpp [hunk 4]
```diff
@@ -11092,6 +11034,105 @@ void HandleTurncoatAttempt( SOLDIERTYPE* pSoldier )
...
 
+void EscapeTimerCallback()
+{
+    const bool chanceToEscape = Chance(75);
+    bool escaped = false;
+    // Look for an escape direction for remaining mercs
+    std::vector<WorldDirections> possibleEscapeDirections{ NORTH, EAST, SOUTH, WEST };
+    std::random_shuffle(possibleEscapeDirections.begin(), possibleEscapeDirections.end());
+
+    for (auto direction : possibleEscapeDirections)
```
Comment (majcosta) [inline Tactical/Overhead.cpp:11045]: I'd const `direction`

File: Tactical/Overhead.cpp [hunk 5]
```diff
@@ -11092,6 +11034,105 @@ void HandleTurncoatAttempt( SOLDIERTYPE* pSoldier )
...
+    {
+        if (IsEscapeDirectionValid(direction) && chanceToEscape && gbWorldSectorZ == 0) // There is no escaping underground! For now..
+        {
+            escaped = true;
+            ScreenMsg(FONT_MCOLOR_LTYELLOW, MSG_INTERFACE, szPrisonerTextStr[STR_PRISONER_ESCAPE]);
+
+            JumpIntoEscapedSector(direction);
+            break;
```
Comment (majcosta) [inline Tactical/Overhead.cpp:11053]: if we're here, escaped has been set to true and there's either no need for this break or for the next if statement

File: Tactical/Overhead.cpp [hunk 6]
```diff
@@ -11092,6 +11034,105 @@ void HandleTurncoatAttempt( SOLDIERTYPE* pSoldier )
...
+    if (!escaped)
+    {
+        ScreenMsg(FONT_MCOLOR_LTYELLOW, MSG_INTERFACE, szPrisonerTextStr[STR_PRISONER_NO_ESCAPE]);
+    }
+}
+
+void AttemptToCapturePlayerSoldiers()
+{
+#ifdef JA2UB
+    ScreenMsg(FONT_MCOLOR_LTYELLOW, MSG_INTERFACE, szPrisonerTextStr[STR_PRISONER_REFUSE_TAKE_PRISONERS]);
```
Comment (majcosta) [inline Tactical/Overhead.cpp:11066]: we probably should just skip the whole surrender codepath if it's UB, no?

File: Strategic/Queen Command.cpp [hunk 7]
```diff
@@ -2716,60 +2716,54 @@ void BeginCaptureSquence( )
...
+int CalculateMaximumPrisonerAmount()
 {
-	UINT32					i;
-	BOOLEAN       fMadeCorpse;
-	INT32         iNumEnemiesInSector;
+	int maxPOWs = 0;
+#ifndef JA2UB
+	if (gubQuest[QUEST_HELD_IN_ALMA] == QUESTNOTSTARTED) { maxPOWs += std::size(gModSettings.iInitialPOWGridNo);  return maxPOWs; }
+	if (gubQuest[QUEST_HELD_IN_TIXA] == QUESTNOTSTARTED) { maxPOWs += std::size(gModSettings.iTixaPrisonPOWGridNo); return maxPOWs; }
+	if (gubQuest[QUEST_INTERROGATION] == QUESTNOTSTARTED) { maxPOWs += std::size(gModSettings.iMeanwhileInterrogatePOWGridNo); return maxPOWs; }
```
Comment (majcosta) [inline Strategic/Queen Command.cpp:2757]: is this supposed to short-circuit like that? If so, I'm guessing captured mercs can't go 30% to one facility, 50% to another and 20% to a third one?

File: Strategic/Queen Command.cpp [hunk 8]
```diff
@@ -2780,6 +2774,7 @@ void EnemyCapturesPlayerSoldier( SOLDIERTYPE *pSoldier )
 	{
 		pSoldier->stats.bLife = 0;
 		pSoldier->iHealableInjury = 0; // added by SANDRO
+		BOOLEAN fMadeCorpse;
```
Comment (majcosta) [inline Strategic/Queen Command.cpp:2777]: hahaha 100% out of scope, but just look at this in HandleSoldierDeath :rofl:

File: Strategic/Queen Command.cpp [hunk 9]
```diff
@@ -2800,32 +2795,32 @@ void EnemyCapturesPlayerSoldier( SOLDIERTYPE *pSoldier )
...
-	ChangeSoldiersAssignment( pSoldier, ASSIGNMENT_POW );
-	RemoveCharacterFromSquads( pSoldier );
-
-	WORLDITEM			WorldItem;
-	std::vector<WORLDITEM> pWorldItem;
-
-#ifdef JA2UB
-	if (gStrategicStatus.ubNumCapturedForRescue < 3 && (gubQuest[QUEST_HELD_IN_ALMA] == QUESTNOTSTARTED || gubQuest[QUEST_INTERROGATION] == QUESTNOTSTARTED))
-#else
+#ifndef JA2UB
```
Comment (majcosta) [inline Strategic/Queen Command.cpp:2804]: It'd be great if UB could be gated in this surrendering functionality in a single spot.

Comment (majcosta) [inline Strategic/Queen Command.cpp:2804]: I thought UB didn't do captures? Any reason this entire function shouldn't be inside this?

File: Strategic/Queen Command.cpp [hunk 10]
```diff
@@ -2883,34 +2878,41 @@ void EnemyCapturesPlayerSoldier( SOLDIERTYPE *pSoldier )
...
-	pSoldier->sBreathRed = 0;
-	pSoldier->flags.fMercCollapsedFlag = FALSE;
+		// make him quite exhausted when found
+		pSoldier->bBreath = pSoldier->bBreathMax = 50;
+		pSoldier->sBreathRed = 0;
+		pSoldier->flags.fMercCollapsedFlag = FALSE;
+
+
+		RemoveSoldierFromTacticalSector(pSoldier, TRUE);
+		extern BOOLEAN RemovePlayerFromTeamSlotGivenMercID(UINT8 ubMercID);
```
Comment (majcosta) [inline Strategic/Queen Command.cpp:2911]: #include "Interface Panels.h"?

File: Tactical/Overhead.cpp (hunk 1 above)
Comment (Asdow) [reply Tactical/Overhead.cpp:11039]: Wouldn't that make it be called only once, and then the chance is set in stone for every subsequent surrender during the whole duration of the program?

File: Tactical/Overhead.cpp (hunk 6 above)
Comment (Asdow) [reply Tactical/Overhead.cpp:11066]: This was the easiest way to deal with offering to surrender ourselves to the enemy. No need to disable or delete the button from UI, enemy simply wants you dead, always.

AI routines and quests related to POW are skipped via compiler switches.

File: Strategic/Quests.h [hunk 11]
```diff
@@ -756,10 +756,13 @@ extern BOOLEAN CheckForNewShipment( void );
...
 extern BOOLEAN CheckTalkerUnpropositionedFemale( void );
 
+enum PowQuestState
+{
+	Q_FAIL,
+	Q_SUCCESS,
+	Q_RESET,
+	Q_END,
+	Q_LEFT_SECTOR
+};
```
Comment (majcosta) [inline Strategic/Quests.h:766]: consider using `enum class` for this if you don't need them to implicity convert to int

File: Strategic/Queen Command.cpp (hunk 7 above)
Comment (Asdow) [reply Strategic/Queen Command.cpp:2757]: The quest gets reset if the player loses control of a prison sector, so it works correctly. 

I should clean this up some though, at first I was aiming on expanding possible POW amounts on prisons, and then knowing how many can currently fit in a prison would've been important. I altered it only slightly with the short circuits when I pivoted back to first getting POW feature simply working.

File: Strategic/strategicmap.cpp [hunk 12]
```diff
@@ -4301,6 +4287,69 @@ void JumpIntoAdjacentSector( UINT8 ubTacticalDirection, UINT8 ubJumpCode, INT32
...
+			}
+			else
+			{
+				if (!AddCharacterToSquad(pSoldier, currentSquad))
+				{
+					currentSquad = AddCharacterToUniqueSquad(pSoldier);
+				}
+			}
+		}
+	}
```
Comment (majcosta) [inline Strategic/strategicmap.cpp:4314]:
```cpp
for (SOLDIERTYPE* pSoldier = MercPtrs[i]; i <= lastID; ++i, ++pSoldier)
{
  // Are we not active in sector
  if ( !pSoldier->bActive || !pSoldier->bInSector || pSoldier->stats.bLife >= OKLIFE) {
    continue;
  }
  if (currentSquad == -1) {
    currentSquad = AddCharacterToUniqueSquad(pSoldier);
    continue;
  }
  if (!AddCharacterToSquad(pSoldier, currentSquad))
  {
    currentSquad = AddCharacterToUniqueSquad(pSoldier);
  }
}

File: Tactical/Overhead.cpp [hunk 13]
```diff
@@ -10298,7 +10231,10 @@ void    DoneFadeOutDueToDeath( )
 void EndBattleWithUnconsciousGuysCallback( UINT8 bExitValue )
 {
     // Enter mapscreen.....
-    if(!is_client)CheckAndHandleUnloadingOfCurrentWorld();
+    if (!is_client)
+    {
+        CheckAndHandleUnloadingOfCurrentWorld();
+    }
```
Comment (majcosta) [inline Tactical/Overhead.cpp:10237]: In the future, avoid mixing whitespace/formatting/etc changes with code changes. I know its tempting and I do it myself, but when reviewing it's balls

File: Tactical/Overhead.cpp (hunk 1 above)
Comment (majcosta) [reply Tactical/Overhead.cpp:11039]: yes, I was wrong there. my bad.

**Comment by majcosta** [comment]
concluding: https://twitter.com/iamdevloper/status/397664295875805184

File: Tactical/Overhead.cpp (hunk 3 above)
Comment (Asdow) [reply Tactical/Overhead.cpp:11043]: Like this?
std::shuffle(possibleEscapeDirections.begin(), possibleEscapeDirections.end(), std::mt19937(time(nullptr)));

Comment (majcosta) [reply Tactical/Overhead.cpp:11043]: https://en.cppreference.com/w/cpp/algorithm/random_shuffle see the examples

File: Strategic/Queen Command.cpp (hunk 8 above)
Comment (Asdow) [reply Strategic/Queen Command.cpp:2777]: -_-

File: Strategic/Quests.h (hunk 11 above)
Comment (Asdow) [reply Strategic/Quests.h:766]: I'm getting identifier "Q_END" is undefined errors in strategicmap.cpp after changing this to enum class

**Review by majcosta — APPROVED** [review]
LGTM :upside_down_face:

**Pull request merged**